### PR TITLE
#273 mutations 5

### PR DESCRIPTION
**Pull request by joshxg**
Bit big!

- Extremely tedious and frustrating without the real api but this adds inserting of invoice lines using the extracted schema......
- It's 'working' for both the mock server and mock worker (persisting, at least)
- There are a bunch of updates still needed, but thought to PR this for now as it's getting too big
- There are values which when added show in the table as `null`/`undefined` - i.e. for `batch` which 1) likely need values added and 2) need to handle this better in cells, probably
- You can only add a single line at a time, at the moment as the `id` given in `ItemDetailsModal` is always `""`, and so when updating/inserting lines we insert with id `""` and then update id `""`. Will add a `uuid` package to generate these in the next one
- Not handling the response right now - it's too difficult to try and handle this meaningfully without the proper api setup

### CODE REVIEW

File: packages/common/src/operations.graphql [hunk 1]
```diff
@@ -529,3 +529,23 @@ query stockCounts {
...
+    __typename
+    insertOutboundShipmentLines {
+      __typename
+      id
+    }
+    updateOutboundShipments {
+      __typename
+      id
+    }
+  }
```
Comment (joshxg) [inline packages/common/src/operations.graphql:550]: not really working with the response right now, so have just left it like this

File: packages/invoices/src/OutboundShipment/DetailView/DetailView.tsx [hunk 2]
```diff
@@ -17,7 +17,7 @@ import { getOutboundShipmentDetailViewApi } from '../../api';
 import { GeneralTab } from './tabs/GeneralTab';
 import { ItemDetailsModal } from './modals/ItemDetailsModal';
 
-import { InvoiceLineRow } from './types';
+import { OutboundShipmentRow } from './types';
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/DetailView.tsx:20]: I changed what we were using as `InvoiceLineRow` to be `OutboundShipmentRow`

File: packages/invoices/src/OutboundShipment/DetailView/modals/ItemDetailsModal.tsx [hunk 3]
```diff
@@ -12,28 +11,28 @@ import {
 
 import { BatchesTable } from './BatchesTable';
 import { ItemDetailsForm } from './ItemDetailsForm';
-import { BatchRow } from '../types';
+import { BatchRow, OutboundShipmentRow } from '../types';
 
 interface ItemDetailsModalProps {
-  invoiceLine?: InvoiceLine;
+  invoiceLine?: OutboundShipmentRow;
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/modals/ItemDetailsModal.tsx:17]: item details modal using `OutboundShipmentRow` over `InvoiceLine` now

File: packages/invoices/src/OutboundShipment/DetailView/modals/ItemDetailsModal.tsx [hunk 4]
```diff
@@ -12,28 +11,28 @@ import {
...
 export const getInvoiceLine = (
   id: string,
   item: Item,
   line: { id: string; expiryDate?: string | null },
   quantity: number
-): InvoiceLine => ({
+): OutboundShipmentRow => ({
   id,
-  itemId: '',
+  itemId: item.id,
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/modals/ItemDetailsModal.tsx:30]: need this for the mutation

File: packages/invoices/src/OutboundShipment/DetailView/reducer.test.ts [hunk 5]
```diff
@@ -2,13 +2,14 @@ import {
   DocumentAction,
   Invoice,
   createColumns,
+  DocumentActionSet,
 } from '@openmsupply-client/common';
 import { placeholderInvoice } from './index';
 import { reducer, OutboundShipmentStateShape, OutboundAction } from './reducer';
-import { InvoiceLineRow } from './types';
+import { OutboundShipmentRow, OutboundShipmentAction } from './types';
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/reducer.test.ts:9]: quite a few test updates to try make them easier to write as I can see quite a few being written!

File: packages/invoices/src/OutboundShipment/DetailView/reducer.test.ts [hunk 6]
```diff
@@ -20,6 +21,11 @@ describe('DetailView reducer', () => {
       costPricePerPack: 0,
       sellPricePerPack: 0,
       itemName: 'a',
+      stockLineId: '',
+      invoiceId: '',
+      isCreated: false,
+      isUpdated: false,
+      isDeleted: false,
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/reducer.test.ts:28]: I've added these `isX` fields to track what mutation to perform for each row. I've used them from the perspective of the client.. as in "isCreated" is `true` when it is created on the client but not persisted on the server, `isDeleted` is true when the row is deleted on the client, but not on the server. I was unsure whether it should be the reverse - i.e. `isCreated` is `true` when it is a persisted row, and when we create one on the client we set that to `false`  😕

File: packages/invoices/src/OutboundShipment/DetailView/reducer.ts [hunk 7]
```diff
@@ -104,20 +104,31 @@ export const reducer = (
...
+            const draftLine = draft.lines.find(
+              line => line.id === serverLine.id
+            );
+
+            if (draftLine) {
+              return mergeLines(serverLine, draftLine);
+            }
+
+            return createLine(serverLine, draft, dispatch);
+          });
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/reducer.ts:121]: Now, after fetching server data, iterate over and if there exists a version of that line in the draft, do a 'merge' - this was primarily done now to make the `isCreated` set to `false` after it had been created.

I don't know if we'd ever be able to set `isUpdated` to false without blocking while saving (i.e. a row `isUpdated=true` -> save -> edit same row again -> save success, don't know what to mark as updated) or having another state field 🤔  - was thinking to just send all lines as updates for now... however, without knowing this we can't determine if the invoice has been fully persisted and show some sort of warning when leaving the page if not.

There's also a chance here:
1. Add a bunch of lines
2. Save
3. Add a new line
4. when merging, only takes the server lines, so newly added line is deleted

Think you can just append any lines in the draft which aren't in the server data, though

File: packages/invoices/src/api.ts [hunk 8]
```diff
@@ -114,7 +118,10 @@ export const onRead =
 
     return {
       ...invoice,
-      lines: linesGuard(invoice.lines),
+      lines: linesGuard(invoice.lines).map(line => ({
+        ...line,
+        stockLineId: '',
```
Comment (joshxg) [inline packages/invoices/src/api.ts:123]: I added `stockLineId` to `InvoiceLine` and it's not on `InvoiceLineNode`.. 🙄  .. but it is on `OutboundShipmentRow` -  sigh! Needs tidying up, just doing what I have to right now 🤣

File: packages/invoices/src/api.ts [hunk 9]
```diff
@@ -134,20 +141,43 @@ const invoiceToInput = (
...
+      if (
+        updateOutboundShipments?.[0]?.__typename ===
+        'UpdateOutboundShipmentResponseWithId'
+      ) {
+        return patch;
+      }
     }
 
-    throw new Error(updateOutboundShipment.error.description);
+    throw new Error(':shrug');
```
Comment (joshxg) [inline packages/invoices/src/api.ts:180]: Want to wait for the real api to handle everything better...

File: packages/mock-server/src/worker/handlers.ts [hunk 10]
```diff
@@ -25,7 +27,7 @@ const insertInvoice = graphql.mutation(
     const { variables } = request;
     const { id, otherPartyId } = variables;
 
-    const result = Api.MutationService.insert.invoice({
+    const result = MutationService.insert.invoice({
```
Comment (joshxg) [inline packages/mock-server/src/worker/handlers.ts:30]: had import problems 🙄

**Review by mark-prins — APPROVED** [review]
🎉  wooo!! another huge effort! thanks 🙇‍♂️ 

I've read through it a few times now and it looks fine to me! Runs against mocks - haven't quite finished getting the rust API working with mutations.

would be good to get it merged!

**Pull request merged**